prusti-viper/src/encoder/encoder.rs

```diff
@@ -344,18 +344,59 @@ impl<'v, 'r, 'a, 'tcx> Encoder<'v, 'r, 'a, 'tcx> {
         args: &Vec<places::Local>,
         target: places::Local,
     ) -> ProcedureContract<'tcx> {
-        let opt_fun_spec = self.get_spec_by_def_id(proc_def_id);
-        let fun_spec = match opt_fun_spec {
-            Some(fun_spec) => fun_spec.clone(),
-            None => {
-                debug!("Procedure {:?} has no specification", proc_def_id);
-                SpecificationSet::Procedure(vec![], vec![])
-            }
+        // get specification on trait declaration method or inherent impl
+        let fun_spec = if let Some(spec) = self.get_spec_by_def_id(proc_def_id) {
+            spec.clone()
+        } else {
+            debug!("Procedure {:?} has no specification", proc_def_id);
+            SpecificationSet::Procedure(vec![], vec![])
         };
+
         let tymap = self.typaram_repl.borrow_mut();
         assert!(tymap.len() == 1);
+
+        // get receiver object base type
+        let tcx = self.env().tcx();
+        let mut impl_spec = SpecificationSet::Procedure(vec![], vec![]);
+        let sig = tcx.fn_sig(proc_def_id);
+
+        if sig.inputs().skip_binder().len() > 0 {
+            let self_ty_binder = sig.input(0);
+            let mut self_ty = self_ty_binder.skip_binder().clone();
+
+
+            // stip reference on self type
+            if let ty::TypeVariants::TyRef(_, true_ty, _) = self_ty.sty {
+                self_ty = true_ty;
+            }
+
+            // get specification on trait impl if exists, and has receiver object
+            if self_ty.is_self() {
+                if let Some(ty) = tymap[0].get(self_ty) {
+                    if let Some(id) = tcx.trait_of_item(proc_def_id) {
+                        let proc_name = tcx.item_name(proc_def_id).to_string();
+                        tcx.for_each_relevant_impl(id, ty, |impl_id| {
+                            for assoc_item in tcx.associated_items(impl_id) {
+                                if assoc_item.name == proc_name {
+                                    if let Some(spec) = self.get_spec_by_def_id(assoc_item.def_id) {
```

Comment:
In one execution of `for_each_relevant_impl`, do you expect this line (`get_spec_by_def_id`) to be executed just once or multiple times (for different `assoc_item` or `impl_id`)?

In the former case, it might be better to split the implementation in two steps:
1. Find the appropriate `assoc_item` with a call to `for_each_relevant_impl` (or with the method suggested in the previous comment).
2. Call `get_spec_by_def_id` on the `assoc_item` that you found.